calyx/src/passes/infer_share.rs

```diff
@@ -0,0 +1,388 @@
+use crate::analysis::{DominatorMap, ReadWriteSet, ShareSet};
+use crate::errors::CalyxResult;
+use crate::ir;
+use crate::ir::traversal::{
+    Action, ConstructVisitor, Named, VisResult, Visitor,
+};
+use crate::ir::CloneName;
+use std::collections::HashSet;
+use std::{rc::Rc};
+
+const BEGIN_ID: &str = "BEGIN_ID";
+const END_ID: &str = "END_ID";
+
+//given an RRC of a cell, determines if cell is a std_mem_cell. There may be an 
+//easier way to do this.
+fn is_std_mem(cell: &ir::Cell) -> bool{
```

Comment:
Why are memories specially defined to be unshareable? They also don't have the `state_share` attribute right?